### PR TITLE
Make KFServer HTTP requests asynchronous

### LINKED ISSUE #789: Transformer request to Predictor should be Non Blocking
/kind feature

**Describe the solution you'd like**
Current requests from the Transformer to the Predictor are blocking: https://github.com/kubeflow/kfserving/blob/49eae18d669222cc6f923f7185c8000c17464f4c/python/kfserving/kfserving/kfmodel.py#L47 This will block the executing Tornado process, leading to a max concurrency through a transformer of the number of Tornado processes, which in the transformer is automatically the number of CPUs in the system. These threads are mostly idle, and blocked, wasting resources. By using an async HTTP call each transformer instance could handle a lot more simultaneous requests: 
https://www.tornadoweb.org/en/stable/guide/async.html

### PR DESCRIPTION
**Pull request by salanki**
**What this PR does / why we need it**:

Requests from KFServer are currently block. This leads to a transformer not being able to make more than number of Tornado workers concurrent requests to the predictor.

As predictors are usually the bottleneck, and transformers are fast, the transformer becomes an unnecessarily bottleneck.

This PR applies Tornado best practices, making HTTP requests async. This drastically improves the transformer performance.

**Before**
Here we scale transformer instances with 40 worker threads each, to handle a total of 300 requests
Screen Shot 2020-04-21 at 14 40 29

**After**
On one transformer instance, with 1 thread. The target number of concurrent requests is much quicker achieved on the predictor, while consuming less resources.
Screen Shot 2020-04-21 at 14 39 08

**Which issue(s) this PR fixes** *
Fixes #789

**Special notes for your reviewer**:

**Release note**:
```release-note
Improve transformer concurrency and performance by making Asynchronous HTTP calls.
```

### CODE REVIEW

File: python/kfserving/kfserving/kfmodel.py [hunk 1]
```diff
@@ -30,6 +32,17 @@ def __init__(self, name: str):
         self.ready = False
         self.predictor_host = None
         self.explainer_host = None
+        # The timeout matches what is set in generated Istio resorurces.
+        # We generally don't want things to time out at the request level here,
+        # timeouts should be handled elsewhere in the system.
+        self.timeout = 600
```
Comment (salanki) [inline python/kfserving/kfserving/kfmodel.py:38]: I wasn't sure if this should be set to 600 or something even larger.

**Comment by yuzisun** [comment]
Thanks for taking on this @salanki !  Can you help sign the CLA?

/ok-to-test

**Comment by salanki** [comment]
I signed the CLA, but it doesn't pick up this E-Mail for some reason. Will force push with a different E-Mail.

**Comment by salanki** [comment]
CLA good and fixed lint.

**Comment by salanki** [comment]
Lint is good. No idea why it says failed.

**Comment by yuzisun** [comment]
looks failing in sdk test  but logs are gone now
http://testing-argo.kubeflow.org/workflows/kubeflow-test-infra/kubeflow-kfserving-presubmit-e2e-791-445bd80-2080-bb45?tab=workflow

/retest

**Comment by yuzisun** [comment]
```
[E 200423 22:00:38 web:1792] Uncaught exception POST /v1/models/TestModel:predict (127.0.0.1)
   HTTPServerRequest(protocol='http', host='127.0.0.1:33497', method='POST', uri='/v1/models/TestModel:predict', version='HTTP/1.1', remote_ip='127.0.0.1')
   Traceback (most recent call last):
     File "/usr/local/lib/python3.6/site-packages/tornado/web.py", line 1703, in _execute
       result = await result
     File "/mnt/test-data-volume/kubeflow-kfserving-presubmit-e2e-791-445bd80-9632-0056/src/kubeflow/kfserving/python/kfserving/kfserving/handlers/http.py", line
44, in post
       response = await model.predict(request)
   TypeError: object dict can't be used in 'await' expression
[E 200423 22:00:38 web:2250] 500 POST /v1/models/TestModel:predict (127.0.0.1) 5.87ms
------------------------------ Captured log call -------------------------------
ERROR    tornado.application:web.py:1792 Uncaught exception POST /v1/models/TestModel:predict (127.0.0.1)
HTTPServerRequest(protocol='http', host='127.0.0.1:33497', method='POST', uri='/v1/models/TestModel:predict', version='HTTP/1.1', remote_ip='127.0.0.1')
Traceback (most recent call last):
 File "/usr/local/lib/python3.6/site-packages/tornado/web.py", line 1703, in _execute
   result = await result
 File "/mnt/test-data-volume/kubeflow-kfserving-presubmit-e2e-791-445bd80-9632-0056/src/kubeflow/kfserving/python/kfserving/kfserving/handlers/http.py", line 44,
in post
   response = await model.predict(request)
TypeError: object dict can't be used in 'await' expression
ERROR    tornado.access:web.py:2250 500 POST /v1/models/TestModel:predict (127.0.0.1) 5.87ms
________________________ TestTFHttpServer.test_explain _________________________
self = <test.test_server.TestTFHttpServer object at 0x7faae5e2f2b0>
http_server_client = <pytest_tornasync.plugin.AsyncHTTPServerClient object at 0x7faae5e2fef0>
   async def test_explain(self, http_server_client):
       resp = await http_server_client.fetch('/v1/models/TestModel:explain',
                                             method="POST",
>                                             body=b'{"instances":[[1,2]]}')
E       tornado.httpclient.HTTPClientError: HTTP 500: Internal Server Error
test_server.py:64: HTTPClientError
----------------------------- Captured stderr call -----------------------------
[E 200423 22:00:39 web:1792] Uncaught exception POST /v1/models/TestModel:explain (127.0.0.1)
   HTTPServerRequest(protocol='http', host='127.0.0.1:41145', method='POST', uri='/v1/models/TestModel:explain', version='HTTP/1.1', remote_ip='127.0.0.1')
   Traceback (most recent call last):
```

**Comment by yuzisun** [comment]
@salanki seems the test here need the fix after your change https://github.com/kubeflow/kfserving/blob/master/python/kfserving/test/test_server.py#L35

**Comment by salanki** [comment]
I’m digging into it

**Comment by yuzisun** [comment]
@salanki any luck on fixing the tests? Thanks for your contribution!

**Comment by salanki** [comment]
@yuzisun: I had some issues getting the entire test suite to run in my setup. I will get it done ASAP. I have 3 more PRs incoming once I can make all tests run. We just went live with a large latency sensitive workload, found some other optimizations around that. Thanks for your patience!

**Comment by yuzisun** [comment]
@salanki We'd like to hear your success story!

/cc @ellis-bigelow @animeshsingh @cliveseldon

**Comment by yuzisun** [comment]
@salanki any luck on getting this passing the test ?

**Comment by yuzisun** [comment]
@salanki Any chance to get this in? we are going to cut 0.4 release soon.
Looks there are a few unit tests to fix
`python/xgbserver/xgbserver/model.py`
`python/sklearnserver/sklearnserver/test_model.py`
`python/pytorchserver/pytorchserver/test_model.py`

File: python/kfserving/kfserving/kfmodel.py [hunk 2]
```diff
@@ -40,30 +52,34 @@ def preprocess(self, request: Dict) -> Dict:
...
-    def explain(self, request: Dict) -> Dict:
+    async def explain(self, request: Dict) -> Dict:
         if self.explainer_host is None:
             raise NotImplementedError
 
-        response = requests.post(
-            EXPLAINER_URL_FORMAT.format(self.explainer_host, self.name),
-            json.dumps(request)
+        response = await self._http_client.fetch(
+            url=PREDICTOR_URL_FORMAT.format(self.explainer_host, self.name),
```
Comment (yuzisun) [inline python/kfserving/kfserving/kfmodel.py:76]: should be `EXPLAINER_URL_FORMAT`

File: python/kfserving/kfserving/kfmodel.py (hunk 1 above)
Comment (yuzisun) [reply python/kfserving/kfserving/kfmodel.py:37]: what's the default timeout?

Comment (salanki) [reply python/kfserving/kfserving/kfmodel.py:37]: Default is 20 seconds. The idea is to set this to the highest timeout in the stack so timeouts don't happen here.

**Comment by salanki** [comment]
The test that is failing seems unrelated: http://testing-argo.kubeflow.org/workflows/kubeflow-test-infra/kubeflow-kfserving-presubmit-e2e-791-9604023-1169-08ec?tab=workflow&nodeId=kubeflow-kfserving-presubmit-e2e-791-9604023-1169-08ec-1389259937

**Comment by yuzisun** [comment]
@salanki here is error in the explainer e2e test

```
[I 200704 15:11:54 anchor_tabular:44] Arr shape ((1, 12),)
[E 200704 15:11:54 web:1792] Uncaught exception POST /v1/models/isvc-explainer-tabular:explain (127.0.0.1)
   HTTPServerRequest(protocol='http', host='isvc-explainer-tabular-explainer-default.kfserving-ci-e2e-test.svc.cluster.local', method='POST', uri='/v1/models/isvc-explainer-tabular:expla
in', version='HTTP/1.1', remote_ip='127.0.0.1')
   Traceback (most recent call last):
     File "/usr/local/lib/python3.7/site-packages/tornado/web.py", line 1703, in _execute
       result = await result
     File "/kfserving/kfserving/handlers/http.py", line 62, in post
       response = await model.explain(request) if inspect.iscoroutinefunction(model.explain) else model.explain(request)
     File "/alibiexplainer/alibiexplainer/explainer.py", line 82, in explain
       explanation = self.wrapper.explain(request["instances"])
     File "/alibiexplainer/alibiexplainer/anchor_tabular.py", line 48, in explain
       if np.argmax(self.predict_fn(arr).shape) == 0:
     File "/alibiexplainer/alibiexplainer/explainer.py", line 74, in _predict_fn
       return np.array(resp["predictions"])
   TypeError: 'coroutine' object is not subscriptable
```

**Comment by yuzisun** [comment]
@salanki I think the issue here is that although `model.explain` is not a coroutine but under the hood it is actually calling `model.predict` which is now changed to coroutine hence complaining the response object.
```
return np.array(resp["predictions"])
   TypeError: 'coroutine' object is not subscriptable
``` 
One way to fix this is to make alibi explain handler to be async as well since it is call the base predict handler which is a coroutine now.

**Comment by salanki** [comment]
Is the alibi explainer the only one failing now? All the python predictors are successful?

**Comment by yuzisun** [comment]
> Is the alibi explainer the only one failing now? All the python predictors are successful?

yes, I think this is the only one which is failing

File: python/alibiexplainer/alibiexplainer/__main__.py [hunk 3]
```diff
@@ -29,6 +30,9 @@
 def main():
     args, extra = parse_args(sys.argv[1:])
     # Pretrained Alibi explainer
+
+    nest_asyncio.apply()
```
Comment (yuzisun) [inline python/alibiexplainer/alibiexplainer/__main__.py:34]: do we still need this line ?

File: python/alibiexplainer/alibiexplainer/explainer.py [hunk 4]
```diff
@@ -70,7 +71,8 @@ def _predict_fn(self, arr: Union[np.ndarray, List]) -> np.ndarray:
                 instances.append(req_data.tolist())
             else:
                 instances.append(req_data)
-        resp = self.predict({"instances": instances})
+        loop = asyncio.get_running_loop()
+        resp = loop.run_until_complete(self.predict({"instances": instances}))
```
Comment (yuzisun) [inline python/alibiexplainer/alibiexplainer/explainer.py:75]: @cliveseldon looks like we could not make the explain call to be async as the `predict_fn` is invoked in alibi itself, maybe it is out of the scope of this PR but I think it is still beneficial to make explain call async.

File: python/alibiexplainer/alibiexplainer/__main__.py (hunk 3 above)
Comment (salanki) [reply python/alibiexplainer/alibiexplainer/__main__.py:34]: Yes. This is required without an upstream change to `alibi` to make it support an async `predictor_fn`.

Comment (salanki) [reply python/alibiexplainer/alibiexplainer/__main__.py:34]: That line is what actually patches `asyncio` to be reentrant.

File: python/alibiexplainer/alibiexplainer/explainer.py (hunk 4 above)
Comment (cliveseldon) [reply python/alibiexplainer/alibiexplainer/explainer.py:75]: Why do we need explain to be async? The explainer will always be a separate Pod that calls the predictor over the network  in batches multiple times until it has enough evidence to make an explanation.

Comment (cliveseldon) [reply python/alibiexplainer/alibiexplainer/explainer.py:75]: The explainer need to run sequential batch requests to the predictor to build up an understanding of the model to create an explanation. It there needs to run those batch requests sequentially. So the explainer predict calls need to be sequential at present I think.

**Comment by yuzisun** [comment]
/lgtm

Will leave @cliveseldon to review and approve

File: python/alibiexplainer/alibiexplainer/explainer.py (hunk 4 above)
Comment (cliveseldon) [reply python/alibiexplainer/alibiexplainer/explainer.py:75]: The predict function is call here: https://github.com/kubeflow/kfserving/blob/99c604e71fa7355434e13798b0c76bba507e91a2/python/alibiexplainer/alibiexplainer/explainer.py#L73

Could we not simply add an `await` there?

**Comment by salanki** [comment]
@ellistarn why close?

**Comment by ellistarn** [comment]
Apologies. I just rewrote history to fix the contributors list. This PR needs a rebase before it can be reopened.

File: python/alibiexplainer/alibiexplainer/explainer.py (hunk 4 above)
Comment (salanki) [reply python/alibiexplainer/alibiexplainer/explainer.py:75]: Explainers do not need to be async, but the `predict` method in `KFModel` is now async.

My initial thought was to throw in `async` and call it a day as well. You can however not `await` without making the method itself an `async def`. I went down the rabbit hole of trying to make the entire explainer async, but the problem is that a reference to `_predict_fn` is passed to your Alibi lib and is called synchronously there. We would need to refactor the Alibi library to support async, which seems unnecessary.

Awaiting synchronously is what I ended up implementing, that way `_predict_fn` is still synchronous.

I'm personally not super impressed with the async implementation in python. They make it very hard to mix async and sync use cases.

**Comment by salanki** [comment]
PR has been rebased on top of master. Please reopen @ellistarn.

**Comment by ellistarn** [comment]
Looks like it simply needs to be recreated. I don't have permission either. I see the error "The ps/async-http branch was force-pushed or recreated".

**Pull request closed without merging**